packages/std/src/ibc.rs

```diff
@@ -0,0 +1,217 @@
+        channel_id: String,
+        data: Binary,
+        timeout_height: IbcTimeoutHeight,
+        timeout_timestamp: u64,
+        version: u64,
+    },
+    /// This will close an existing channel that is owned by this contract.
+    /// Port is auto-assigned to the contracts' ibc port
+    CloseChannel { channel_id: String },
+}
+
+/// These are queries to the various IBC modules to see the state of the contract's
+/// IBC connection. These will return errors if the contract is not "ibc enabled"
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum IbcQuery {
+    /// Gets the Port ID the current contract is bound to.
+    /// Returns PortIdResponse
+    PortId {},
+    /// Lists all (portID, channelID) pairs that are bound to a given port
+    /// If port_id is omitted, list all channels bound to the contract's port.
+    /// Returns ListChannelsResponse.
+    ListChannels { port_id: Option<String> },
+    /// Lists all information for a (portID, channelID) pair.
+    /// If port_id is omitted, it will default to the contract's own channel.
+    /// (To save a PortId{} call)
+    /// Returns ChannelResponse.
+    Channel {
+        channel_id: String,
+        port_id: Option<String>,
+    },
+    // TODO: Add more
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct PortIdResponse {
+    pub port_id: String,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct ListChannelsResponse {
+    pub channels: Vec<IbcEndpoint>,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct ChannelResponse {
+    pub channel: IbcChannel,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct IbcEndpoint {
+    pub port_id: String,
+    pub channel_id: String,
+}
+
+// These are various messages used in the callbacks
+
+/// IbcChannel defines all information on a channel.
+/// This is generally used in the hand-shake process, but can be queried directly.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct IbcChannel {
+    pub endpoint: IbcEndpoint,
+    pub counterparty_endpoint: IbcEndpoint,
+    pub order: IbcOrder,
+    pub version: String,
+    /// CounterpartyVersion can be None when not known this context, yet
+    pub counterparty_version: Option<String>,
+    /// The connection upon which this channel was created. If this is a multi-hop
+    /// channel, we only expose the first hop.
+    pub connection_id: String,
+}
+
+// TODO: check what representation we want here for encoding - string or number
+/// IbcOrder defines if a channel is ORDERED or UNORDERED
+/// Values come from https://github.com/cosmos/cosmos-sdk/blob/v0.40.0/proto/ibc/core/channel/v1/channel.proto#L69-L80
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub enum IbcOrder {
+    Unordered = 1,
+    Ordered = 2,
```

Comment:
The numeric values get lost in the json serialization. Only the field names are serialized. See https://github.com/CosmWasm/serde-json-wasm/pull/23

So we can remove the numbers and pick up the strings in the Go code.

Reply:
I would leave them here as reference. I want to see how this sorts out on the go side.

Reply:
This is how the naming works on the go side: https://github.com/cosmos/cosmos-sdk/blob/master/x/ibc/core/04-channel/types/channel.pb.go#L83-L93

It may end up as "ORDER_UNORDERED" by default. I would rather have the info here (even if it changes nothing) and then sort it out when we tie into the sdk

Reply:
But the numeric values will never reach the CosmWasm VM, and not the Go side. I think you should never write explicit enum values if there is not very good reason to do so, i.e. their numeric value is used. Then it should be documented to which other system the numers need to match.

In order to use the strings you need in Go, better annotate as follows:

```rust
pub enum IbcOrder {
     #[serde(rename = "ORDER_UNORDERED")]
    Unordered,
     #[serde(rename = "ORDER_ORDERED")]
    Ordered,
}
```

Reply:
Done